ChangeLog.d/psa_alg_tls12_ecjpake_to_pms-reject_ka.txt

```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * The key derivation algorithm PSA_ALG_TLS12_ECJPAKE_TO_PMS cannot be
+     used on a shared secret from a key agreement since its input must be
```

Comment:
I find this weird. Why does the *secret* input of a key derivation need to have one specific format which is meant to be a public key?

Reply:
Because this "KDF" is actually is really ad-hoc. The PMS is defined as SHA-256(K.X) where K is the shared secret from the EC J-PAKE key exchange, and K.X is it's X coordinate. Since the PSA EC J-PAKE API passes the serialization of K to the KDF, what this KDF needs to do is extract bytes 1-32 included, feed them to SHA-256, and output the result.

So, the KDF as the very least needs to check that it input is at least 33 bytes, but it seemed natural to actually assert that the input is exactly 65 bytes and starts with 0x04, but if that's annoying I guess we can remove it.

Btw, I think the sentence is incorrect: "cannot be used on a shared secret from a key agreement _other than_ EC J-PAKE with secp256r1". It can absolutely be used with that particular key exchange, because that's what we do in the implementation of the ECJPAKE TLS 1.2 key exchange when `USE_PSA_CRYPTO` is enabled.

Reply:
ECJPAKE isn't a key agreement as far as the PSA API is concerned (it's not something you can do with `psa_raw_key_agreement` or `psa_key_agreement_key_derivation`). The output of an ECDH key agreement on a Weierstrass curve is the X coordinate in big-endian order. Why doesn't `PSA_ALG_TLS12_ECJPAKE_TO_PMS` take that as its input?

Reply:
> ECJPAKE isn't a key agreement as far as the PSA API is concerned 

Ah, of course, good point.

> The output of an ECDH key agreement on a Weierstrass curve is the X coordinate in big-endian order. Why doesn't `PSA_ALG_TLS12_ECJPAKE_TO_PMS` take that as its input?

Because it's designed to be use with `psa_pake_get_implicit_key()` which according to [its documentation](https://arm-software.github.io/psa-api/crypto/1.1/ext-pake/api/pake.html#c.psa_pake_get_implicit_key) " injects the shared secret output of the PAKE into the provided key derivation operation" and for [JPAKE](https://arm-software.github.io/psa-api/crypto/1.1/ext-pake/api/pake.html#c.PSA_ALG_JPAKE) that's a group element which for [ECC](https://arm-software.github.io/psa-api/crypto/1.1/ext-pake/api/pake.html#c.PSA_PAKE_PRIMITIVE_TYPE_ECC) uses "the same [format] as that for public keys in the specific Diffie-Hellman group. For more information, consult the documentation of psa_export_public_key()."

So, `PSA_ALG_TLS12_ECJPAKE_TO_PMS` take an ECC point in the standard format because that's what `psa_pake_get_implicit_key()` will give it. We further restricted it to secp256r1 (or any other 256-bit short Weierstrass curve)  because that's the only thing that's specified for use in TLS ECJPAKE anyway. But the format (04 | X | Y) is given by the PSA spec.

Reply:
> "injects the shared secret output of the PAKE into the provided key derivation operation" and for JPAKE that's a group element which for ECC uses "the same [format] as that for public keys in the specific Diffie-Hellman group".

But why? Why not use the X coordinate for the shared secret?

(That's why as in why the spec was written that way, so “because the spec says so” is not an answer. The spec is beta, we can change it if we notice difficulties during the implementation.)

Reply:
It would be interesting to know what other protocols based on EC J-PAKE are doing and how what exactly they pass to the KDF. If that's always X, or something that can be derived from X, then we can indeed consider passing only X.

Reply:
> Then I think that's a question for @yanesca but I'd say that's probably because it's the most generic format and contains the full information.

Indeed, I choose that because it seemed the most universal and conceptually the simplest solution. I admit that it is unlikely that the Y coordinate is needed, but if it is, then it is more complicated to derive it than just checking a prefix and discarding trailing bytes.

Reply:
Note: as noted [elsewhere](https://github.com/ARM-software/psa-crypto-api/issues/564#issuecomment-1328827440) there's an ISO standard about J-PAKE which seems suggest a KDF to use with it. If that's the case, hopefully it also has something to say about what the input to the KDF should be. Unfortunately we don't seem to have access to a copy of that standard atm.